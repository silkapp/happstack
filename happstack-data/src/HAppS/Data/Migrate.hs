{-# LANGUAGE MultiParamTypeClasses #-}
module HAppS.Data.Migrate where

class Migrate a b where
    migrate :: a -> b

