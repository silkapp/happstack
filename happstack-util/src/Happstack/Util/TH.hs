{-# LANGUAGE TemplateHaskell #-}
module Happstack.Util.TH where

import Language.Haskell.TH

instanceD' :: CxtQ -> TypeQ -> Q [Dec] -> DecQ
instanceD' ctxt ty decs =
    do decs' <- decs
       let decs'' = filter (not . isSigD) decs'
       instanceD ctxt ty (map return decs'')

isSigD :: Dec -> Bool
isSigD (SigD _ _) = True
isSigD _ = False

