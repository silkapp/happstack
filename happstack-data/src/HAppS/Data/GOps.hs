module HAppS.Data.GOps where
import Data.Generics hiding (GT)
import Control.Monad.Reader
import Control.Monad.State
import Data.Maybe

-- useful generic functions with better names

gSet :: (Data b, Typeable a) => a -> b -> b
gSet v = gReplace (const v)

gReplace :: (Typeable a, Data b) => (a -> a) -> b -> b
gReplace f x = everywhere (mkT f) x

gFind :: (MonadPlus m, Data a, Typeable b) => a -> m b
gFind x = msum $ map return $ listify (const True) x

gFind' :: (Data a, Typeable b) => a -> b
gFind' x = fromJust $ gFind x
--Monad versions

gModify :: (MonadState s m,Typeable a,Data s) => (a->a) -> m ()
gModify f = modify $ gReplace f

gAsk :: (Data r, Typeable a, MonadReader r m, MonadPlus n) =>
        (a -> n b) -> m (n b)
gAsk f = do st <- ask
            let y = gFind st 
            return $ maybe mzero f y

gGet :: (Data s, Typeable a, MonadState s m, MonadPlus n) =>
        (a -> n b) -> c -> m (n b)
gGet f _ = do st <- get
              let y = gFind st 
              return $ maybe mzero f y

