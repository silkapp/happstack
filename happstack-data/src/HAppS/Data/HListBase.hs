{-# LANGUAGE TemplateHaskell, FlexibleInstances,
             UndecidableInstances, OverlappingInstances,
             MultiParamTypeClasses, CPP, DeriveDataTypeable #-}
module HAppS.Data.HListBase where
import HAppS.Data.DeriveAll
import HAppS.Data.Default
import Data.Typeable

$( deriveAll [''Show,''Default,''Eq,''Read,''Ord]
   [d|
        data Couple a b = Couple a b
        data Nil = Nil
    |]
  )

nil::Nil
nil=Nil
