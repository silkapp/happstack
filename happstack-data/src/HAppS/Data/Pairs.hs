{-# LANGUAGE TemplateHaskell, FlexibleInstances, UndecidableInstances -- just for example at the bottom
    ,CPP, DeriveDataTypeable, MultiParamTypeClasses
  #-}
module HAppS.Data.Pairs (pairsToXml,xmlToPairs,pairsToHTMLForm,xmlToHTMLForm
                        ,toPairs,toPairsX,fromPairs,toHTMLForm
                        ,Pairs,AsPairs
                        ) where

import Data.Char
import Data.List
import Data.Maybe

---stuff for examples
import HAppS.Data.DeriveAll
import HAppS.Util.Common

import Data.Generics as G
import HAppS.Data.Default -- for pairs
import HAppS.Data.Xml
import Control.Monad.Identity

type Pairs = [(String,String)]

pairsToXml :: Pairs -> [Element]
pairsToXml pairs = fst $ formIntoEls "" $ map slash pairs

slash :: (String,t) -> (String,t)
slash p@('/':_,_) = p
slash (n,v) = ('/':n,v)

formIntoEls :: String -> Pairs -> ([Element], Pairs)
formIntoEls _ [] = ([],[])
formIntoEls ctx pairs@((name,val):rest)
    | not $ isPrefixOf ctx name = ([],pairs)
    | isAttr = moreCtx $ Attr elName val
    | isLeaf = moreCtx $ Elem elName [CData val]
    | otherwise =
        let (es,pairs') = formIntoEls ctx' pairs
            (es',pairs'') = formIntoEls ctx pairs'
        in
        (Elem elName es:es',pairs'')
    where
    ctx' = ctx ++ "/" ++ top
    node =  tail $ drop (length ctx) name
    (top,subs) = break (=='/') $  node
    elName = takeWhile (/='[') top
    isLeaf = null subs
    isAttr = head top == '@'
    moreCtx el = let (restCtx,restPairs) = formIntoEls ctx rest
                 in (el:restCtx,restPairs)
xmlToPairs :: [Element] -> Pairs
xmlToPairs =
    map (\(x,y)->(tail x,y)) .
    xmlIntoPairs 0 ""

xmlIntoPairs :: Int -> String -> [Element] -> Pairs
xmlIntoPairs _ _ [] = []
xmlIntoPairs x ctx (Elem _ []:xs) = xmlIntoPairs x ctx xs
xmlIntoPairs x ctx (Attr n v:xs) = (ctx++"/"++n,v):xmlIntoPairs x ctx xs
xmlIntoPairs _ ctx (CData v:[]) = [(ctx,v)]
xmlIntoPairs _ _ (CData _:_) = []
xmlIntoPairs i ctx ((Elem n xs):xs') =
    thisElPairs ++ restPairs
    where
    --if we have two elements with the same name then we have to add [i]
    nIndex = n ++ "[" ++ (show i) ++ "]"
    (iNext,nName)
        | i/=0 = (i+1,nIndex)
        | null xs' = (0,n)
        | next==n = (i+1,nIndex)
        | otherwise = (0,n)
    Elem next _ = head xs'
    thisElPairs = (xmlIntoPairs 0 (ctx++"/"++nName) xs)
    restPairs = (xmlIntoPairs iNext ctx xs')

pairsToHTMLForm :: String -> String -> String -> Pairs -> [Element]
pairsToHTMLForm iden action method pairs
 = [Elem "form" (Attr "action" action :
                Attr "id" iden :
                Attr "method" method :
                map pToInput pairs ++
                [submitButton])]

submitButton :: Element
submitButton = Elem "input" [Attr "type" "submit"]

pToInput :: (String,String) -> Element
pToInput (n,v)=
    Elem "div" [Attr "class" "formEl",
                Elem "span" [Attr "class" "name"
                            ,CData $ map (\x->if x=='/' then ' ' else x) n]
               ,Elem "input" [Attr "name" n,Attr "value" v]]

xmlToHTMLForm :: (Xml a, Show a, Data a, Eq a) =>
                 String -> String -> String -> a -> [Element]
xmlToHTMLForm iden method action
 = pairsToHTMLForm iden method action  . toPairsX -- xmlToPairs

class (Xml x,Show x, G.Data x) => AsPairs x where
    toPairs::x->Pairs
    fromPairs::Pairs -> Maybe x

instance (Xml a,Show a,G.Data a,Eq a) => AsPairs a where
    toPairs x = xmlToPairs $ toPublicXml x
    fromPairs [] = Nothing
    fromPairs pairs = if res == dv && notRigidMatch then Nothing else Just res
        where
        xml = pairsToXml $ mapFst clean pairs
        res = runIdentity $ fromXml Flexible xml
        mbRigidMatch = fromXml Rigid xml
        _ = [mbRigidMatch,Just res]
        isRigidMatch = isJust mbRigidMatch
        notRigidMatch = not isRigidMatch
        dv = defaultValue
        (cons,_) = break (==' ') $ show dv
        clean n = if (map toLower parent)==(map toLower cons) then n
                  else if head n =='/' then n
                  else (cons++('/':name))
            where
            name = trimSlash n
            (parent,_) = break (=='/') name
        trimSlash n = if head n=='/' then tail n else n

toPairsX :: (Xml a, Show a, Data a, Eq a) => a -> Pairs
toPairsX x = map (\(n,v)->let (_,child)=break (=='/') n in
                            if null child then (n,v) else (tail child,v)) $ toPairs x

toHTMLForm :: (Xml a, Show a, Data a, Eq a) =>
              String -> String -> String -> a -> [Element]
toHTMLForm iden method action = xmlToHTMLForm iden method action 



--- example usage and tests here
$( deriveAll [''Show,''Default,''Eq]
   [d|
       data UserInfo = UserInfo User Pass 
       newtype User = User String 
       newtype Pass = Pass String 
    |]
 )
