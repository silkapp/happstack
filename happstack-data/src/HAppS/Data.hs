module HAppS.Data
   (
    module Data.Typeable
   ,module HAppS.Data.Default
   ,module HAppS.Data.DeriveAll
   ,module HAppS.Data.HList
   ,module HAppS.Data.Migrate
   ,module HAppS.Data.Normalize
   ,module HAppS.Data.Pairs
   ,module HAppS.Data.Xml
   ,module HAppS.Data.GOps
   ,module HAppS.Data.Serialize
   ,module HAppS.Data.SerializeTH
   )
where

import HAppS.Data.GOps
import Data.Typeable
import HAppS.Data.Default
import HAppS.Data.DeriveAll
import HAppS.Data.HList
import HAppS.Data.Migrate
import HAppS.Data.Normalize
import HAppS.Data.Pairs
import HAppS.Data.Xml
import HAppS.Data.Serialize hiding (migrate, Migrate)
import HAppS.Data.SerializeTH
