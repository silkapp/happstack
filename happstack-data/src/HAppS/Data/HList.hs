{-# LANGUAGE TemplateHaskell, FlexibleInstances,
             UndecidableInstances, OverlappingInstances,
             MultiParamTypeClasses, CPP, FunctionalDependencies #-}

module HAppS.Data.HList (HasT, hlextract, hlupdate, (.&.),
                         (:&:),
                         Couple(..),Nil(..),CoupleClass,hMap,trans) where
-- HList useful with generic

import HAppS.Data.Xml
import HAppS.Data.Pairs
import Data.Generics as G
import HAppS.Data.HListBase

infixr 6 .&.
(.&.) :: a -> b -> Couple a b
(.&.) = Couple

type a :&: b = Couple b a



class CoupleClass a where
    toPairs' :: a -> Pairs
    fromPairs' :: Pairs -> Maybe a

instance (Eq a,Xml a, Show a, G.Data a,CoupleClass b) => CoupleClass (Couple a b) where
    toPairs' (Couple a b) = (toPairs a) ++ (toPairs' b)
    fromPairs' = const Nothing --doesn't make sense to unmix to make toPairs' . fromPairs' an identity

instance CoupleClass Nil where
    toPairs' _ = []
    fromPairs' _ = return Nil

instance (Xml a, Xml b) => Xml (Couple a b) where
    toXml (Couple a b) = (toXml a) ++ (toXml b)
    readXml r xml = do
                  (xml', a) <- readXml r xml
                  (xml'', b) <- readXml r xml'
                  return (xml'', Couple a b)

hlextract :: HasT a b => a -> b
hlextract hlist = x hlist

hlupdate :: HasT a b => a -> b -> a
hlupdate hlist val = u hlist val

class HasT a b where
    x :: a -> b
    u :: a -> b -> a

instance HasT (Couple a b) a  where
    x (Couple a _) = a
    u (Couple _ b) a = Couple a b

instance HasT (Couple a b) b where
    x (Couple _ b) = b
    u (Couple a _) b = Couple a b

-- Oleg's trick http://www.haskell.org/pipermail/haskell/2004-June/014176.html
class HasT' a b where 
    x' :: a -> b
    u' :: a -> b -> a

instance HasT' a b => HasT a b where
    x a = x' a
    u a b = u' a b

instance (HasT c a) => HasT' (Couple b c) a where
    x' (Couple _ b) = x b
    u' (Couple a b) c = Couple a (u b c)

class Trans ft a where
    trans :: ft -> a -> a

instance Trans (a->a) (Couple a b) where
    trans f (Couple a b) = Couple (f a) b

instance Trans (b->b) (Couple a b) where
    trans f (Couple a b) = Couple a (f b) 

class Trans' ft a where
    trans' :: ft -> a ->a

instance Trans' ft a => Trans ft a where
    trans f a = trans' f a

instance (Trans ft b) => Trans' ft (Couple a b) where
    trans' f (Couple a b) = Couple a (trans f b)

class HMap a b | a -> b where
    hMap::a->b

instance (HMap b d,CoupleClass b) => HMap (Couple a b) (Couple [a] d) where
    hMap (Couple a b) = Couple [a] $ hMap b 

instance HMap (Couple a Nil) (Couple [a] Nil) where
    hMap (Couple a Nil) = Couple [a] Nil
