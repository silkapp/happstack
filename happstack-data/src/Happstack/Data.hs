module Happstack.Data
   (
    module Data.Typeable
   ,module Happstack.Data.Default
   ,module Happstack.Data.DeriveAll
   ,module Happstack.Data.Migrate
   ,module Happstack.Data.Normalize
   ,module Happstack.Data.Pairs
   ,module Happstack.Data.Xml
   ,module Happstack.Data.GOps
   ,module Happstack.Data.Serialize
   ,module Happstack.Data.SerializeTH
   )
where

import Happstack.Data.GOps
import Data.Typeable
import Happstack.Data.Default
import Happstack.Data.DeriveAll
import Happstack.Data.Migrate
import Happstack.Data.Normalize
import Happstack.Data.Pairs
import Happstack.Data.Xml
import Happstack.Data.Serialize hiding (migrate, Migrate)
import Happstack.Data.SerializeTH

