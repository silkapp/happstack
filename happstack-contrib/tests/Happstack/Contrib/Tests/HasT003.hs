
{-# LANGUAGE TemplateHaskell, DeriveDataTypeable,
             FlexibleInstances, MultiParamTypeClasses,
             OverlappingInstances, UndecidableInstances #-}

module Main (main) where

import Language.Haskell.TH
import Happstack.Data

$( deriveAll [''Show,''Eq, ''Default]
             [d| newtype User = User String |]
 )

tHasT :: HasT hlist y => hlist -> y -> (y, y)
tHasT hlist v = (hlextract hlist, v)

e9 = tHasT (User "alex" .&. 456) 123

main :: IO ()
main = print e9

