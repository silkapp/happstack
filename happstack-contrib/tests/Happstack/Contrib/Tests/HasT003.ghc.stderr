
HasT003.hs:18:5:
    Overlapping instances for HasT (Couple User t) t1
      arising from a use of `tHasT' at HasT003.hs:18:5-35
    Matching instances:
      instance [overlap ok] (Happstack.Data.HList.HasT' a b) => HasT a b
        -- Defined in Happstack.Data.HList
      instance [overlap ok] HasT (Couple a b) a
        -- Defined in Happstack.Data.HList
      instance [overlap ok] HasT (Couple a b) b
        -- Defined in Happstack.Data.HList
    (The choice depends on the instantiation of `t, t1'
     To pick the first instance above, use -fallow-incoherent-instances
     when compiling the other instance declarations)
    In the expression: tHasT (User "alex" .&. 456) 123
    In the definition of `e9': e9 = tHasT (User "alex" .&. 456) 123
